packages/@aws-cdk/aws-codepipeline/lib/jenkins-provider.ts

```diff
@@ -0,0 +1,256 @@
+      JenkinsTestAction {
+    return new JenkinsTestAction(this, name, {
+      stage,
+      jenkinsProvider: this,
+      ...props,
+    });
+  }
+
+  public abstract _includeBuildVariant(): void;
+  public abstract _includeTestVariant(): void;
+}
+
+/**
+ * A class representing Jenkins providers.
+ *
+ * @see #import
+ */
+export class JenkinsProvider extends BaseJenkinsProvider {
+  /**
+   * Import a Jenkins provider registered either outside the CDK,
+   * or in a different CDK Stack.
+   *
+   * @param parent the parent Construct for the new provider
+   * @param id the identifier of the new provider Construct
+   * @param props the properties used to identify the existing provider
+   * @returns a new Construct representing a reference to an existing Jenkins provider
+   */
+  public static import(parent: cdk.Construct, id: string, props: ImportJenkinsProviderProps): IJenkinsProvider {
+    return new ImportedJenkinsProvider(parent, id, props);
+  }
+
+  public readonly providerName: string;
+  public readonly serverUrl: string;
+  private buildIncluded = false;
+  private testIncluded = false;
+
+  constructor(parent: cdk.Construct, id: string, props: JenkinsProviderProps) {
+    super(parent, id, props.version);
+
+    this.providerName = props.providerName;
+    this.serverUrl = props.serverUrl;
+
+    if (props.forBuild === true) {
+      this._includeBuildVariant();
+    }
+    if (props.forTest === true) {
+      this._includeTestVariant();
+    }
+  }
+
+  public _includeBuildVariant(): void {
+    if (this.buildIncluded) {
+      return;
+    }
+    this.buildIncluded = true;
+    this.registerJenkinsCustomAction('JenkinsBuildProviderResource', cpapi.ActionCategory.Build);
+  }
+
+  public _includeTestVariant(): void {
+    if (this.testIncluded) {
+      return;
+    }
+    this.testIncluded = true;
+    this.registerJenkinsCustomAction('JenkinsTestProviderResource', cpapi.ActionCategory.Test);
+  }
+
+  private registerJenkinsCustomAction(id: string, category: cpapi.ActionCategory) {
+    new CustomActionRegistration(this, id, {
+      category,
+      artifactBounds: jenkinsArtifactsBounds,
+      provider: this.providerName,
+      version: this.version,
+      entityUrl: appendToUrl(this.serverUrl, 'job/{Config:ProjectName}'),
+      executionUrl: appendToUrl(this.serverUrl, 'job/{Config:ProjectName}/{ExternalExecutionId}'),
+      actionProperties: [
+        {
+          name: 'ProjectName',
+          required: true,
+          key: true,
+          queryable: true,
+        },
+      ],
+    });
+  }
+}
+
+class ImportedJenkinsProvider extends BaseJenkinsProvider {
+  public readonly providerName: string;
+  public readonly serverUrl: string;
+
+  constructor(parent: cdk.Construct, id: string, props: ImportJenkinsProviderProps) {
+    super(parent, id, props.version);
+
+    this.providerName = props.providerName;
+    this.serverUrl = props.serverUrl;
+  }
+
+  public _includeBuildVariant(): void {
+    // do nothing
```

Comment:
If we make the forBuild and forTest arguments to the imported props, we can do a sanity check and throw a useful error here.

Reply:
Won't that make using the imported provider more irritating? I assume that `forBuild` and `forTest` will be `false` by default, so if you do:

```typescript
const provider = codepipeline.JenkinsProvider.import(this, 'Provider', {
  providerName: 'Provider',
  serverUrl: 'http://my-jenkins.com:8080',
});

provider.addToPipeline(/* ... */);
```

and you would get an exception saying `forBuild` should be `true` when doing the import.

Not sure it's worth it...?

Reply:
I'm not sure I'm *optimizing* for it, just pointing out a potential usability issue with that use case :)

Reply:
The way I'm looking at it, by deciding whether or not to add that parameter to the ImportProps, we're optimizing for usability of one of the two cases. We just have to decide which one we care more about.

Reply:
I think I prefer the current approach, for 2 reasons:

a) the link between specifying the `forBuild: true` parameter to `import` before calling `provider.addToPipeline(...)` seems very unintuitive to me. I understand we can have a nice error message, but still, I don't love this approach.
b) I can't see a good use case for creating the provider in one Stack, but adding it to the Pipeline in another. Even if someone tried to do it, it wouldn't actually work! The imported Provider would throw an exception saying it wasn't marked for build. Of course, you could pass `forBuild: true` when creating it in the original Stack, but again, I don't find that intuitive.

Personally, if you feel strongly about having this validation, then I would rather go back to having separate Providers for build & test. I don't think forcing passing `forBuild/forTest: true`, and in a call before you get the actual exception, is a great API.

Reply:
Alrighty then. That was the last point of discussion, right? So ship it?

Reply:
> So ship it?

If you don't have any other comments, then yep :)

Reply:
Already approved on my part so go ahead